### PR TITLE
Mhr Filing Submission

### PR DESCRIPTION
**Pull request by garyfan1**
*Issue #:* /bcgov/entity#13215

*Description of changes:*
- added a interface that matches the required properties in the API
- added a composable for constructing the request payload (we do not want to include `''`, `null`, or `undefined` values)
- some validation flags are set for testing purpose
- some properties are set for testing purpose
---
### Testing
- open up the console when clicking on `submit` to see the result
- use 83000007 or above for document id when testing, I increment 1 for every successful submission to avoid conflicts.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue
Comment (dimak1) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue:403]: When this should be removed?

File: ppr-ui/tests/unit/test-data/mock-mhr-registration.ts
Comment (dimak1) [inline ppr-ui/tests/unit/test-data/mock-mhr-registration.ts]: Remove?

File: ppr-ui/tests/unit/test-data/mock-mhr-registration.ts
Comment (dimak1) [inline ppr-ui/tests/unit/test-data/mock-mhr-registration.ts]: Remove?

File: ppr-ui/src/utils/mhr-api-helper.ts
Comment (dimak1) [inline ppr-ui/src/utils/mhr-api-helper.ts:182]: Throwing `Invalid API response` error above would be caught here. Intentional?

File: ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts
Comment (dimak1) [inline ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts]: Remove?

**Review by dimak1 — CHANGES_REQUESTED** [review]
There are bunch of _TODO_ comments to remove the code. Maybe you can clean up? Or when and who suppose to remove it?

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue
Comment (garyfan1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue:403]: it'll be removed after QA

File: ppr-ui/src/utils/mhr-api-helper.ts
Comment (garyfan1) [reply ppr-ui/src/utils/mhr-api-helper.ts:182]: that's what we did in `searchMhrPDF`, I wanted to throw an actual `Error` instead of just returning a object with a property called error.

**Comment by garyfan1** [comment]
The TODOs have to be there to successfully file a submission, there's no way to meet the minimal payload requirement through UI at this stage.

**Comment by dimak1** [comment]
> The TODOs have to be there to successfully file a submission, there's no way to meet the minimal payload requirement through UI at this stage.

Ok fair enough. Could we then split the code into two PRs? One is clean and the other one is with the changes for QA. So we'll know what exactly to revert.

**Comment by cameron-eyds** [comment]
> > The TODOs have to be there to successfully file a submission, there's no way to meet the minimal payload requirement through UI at this stage.
> 
> Ok fair enough. Could we then split the code into two PRs? One is clean and the other one is with the changes for QA. So we'll know what exactly to revert.

@garyfan1 @dimak1 

I'm not sure that extra work is required. Same code in the same place, at the end of the day. 
IF these TODOs are required to remain in the code for QA purposes until some pieces are completed, perhaps update your generic todos to something specific to this commit, ie `// TODO: Mhr-Submission - ...` so they are easily searchable. 

Also to that point, create a ticket in ZENHUB with the task of removing and/or verifying these have been removed as we complete the missing pieces called out by these comments. 
Post a link to that ticket here when you've created it 👍 

* To add to this, you've got to ask the question of IF these are required to remain in. It seems like it will be useful for the purpose of creating/testing outputs as well as moving some blocked API tickets along, so i think* it makes sense. 
Yours guys thoughts on this?

File: ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts
Comment (cameron-eyds) [inline ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts]: Why a function vs a function expression here?

Also a function like this deserves some comments and explanation to help other developers understand what this function is doing. 

Please also break the conditional blocks to make for easier readability using `{}`

File: ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts
Comment (cameron-eyds) [inline ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts]: To be transparent, this isn't how i would have done it. ie removing unwanted props vs constructing something new out of what WAS wanted. 
But also, i don't mind it at all, the only risk i can think of is unwanted properties in the payload. 
ie a developer adds a new property to store or something similar that doesn't have any handling here. 

I suppose it goes both ways since it would be included in the payload without additional work but it could also be included unintentionally. 

Either way, i'm fine with this! Just my thoughts no action required.

File: ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts
Comment (cameron-eyds) [inline ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts:52]: How close is this to our UI equivalent IF?

File: ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts
Comment (cameron-eyds) [inline ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts]: Is this property meant to be commented out? 
In development?

File: ppr-ui/src/store/getters/state-getters.ts
Comment (cameron-eyds) [inline ppr-ui/src/store/getters/state-getters.ts]: new line for consistency.

File: ppr-ui/src/store/state/state-model.ts
Comment (cameron-eyds) [inline ppr-ui/src/store/state/state-model.ts:111]: There is some confusion with the type on these properties. 
In another PR there is work to enforce the type on these as `number`.

Unfortunately it DOES matter, ie string is initialized as this, `number` as `null`. 
So i believe this is right, according to the spec.

File: ppr-ui/src/utils/mhr-api-helper.ts
Comment (cameron-eyds) [inline ppr-ui/src/utils/mhr-api-helper.ts]: new line

File: ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue
Comment (cameron-eyds) [inline ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue]: remove dev code

File: ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts
Comment (garyfan1) [reply ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts]: Because I couldn't figure out the syntax of function expression with type template...

will add some comments.

**Review by cameron-eyds — COMMENTED** [review]
Looks really good, great job on composable. 
Just a few minor comments and as well to Dimas point, we need a means of tracking these todo items that are in the sake of unblocking things. 
Please create/post here a ticket link, that will handle that tech debt.

File: ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts
Comment (garyfan1) [reply ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts]: hmm I think you're making more sense and I think will have less code, I'll look into it

Comment (cameron-eyds) [reply ppr-ui/src/composables/mhrRegistration/useNewMhrRegistration.ts]: No action required from me but if you feel there is a better way by all means.
100% your call.

File: ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts
Comment (garyfan1) [reply ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts:52]: pretty close, mainly just excluding `owners` and the required properties are different

oh and some types are different

File: ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts
Comment (garyfan1) [reply ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts]: it can be removed, especially the second line, kept it there to let people know why this isn't even optional.

File: ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue
Comment (garyfan1) [reply ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue]: I kept it for QA testing.

File: ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts
Comment (cameron-eyds) [reply ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts:52]: Cool, thanks!

File: ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue
Comment (cameron-eyds) [reply ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue]: I don't think QA will need this. 
If a developer wants to see the `result` in the browser, you can get it from the network request response.

File: ppr-ui/src/store/state/state-model.ts
Comment (ChasBoucher) [reply ppr-ui/src/store/state/state-model.ts:111]: What happens when you test the api call? As Cameron was saying, there is another ticket that ensures it is a number(I am working on that one)...

**Review by ChasBoucher — COMMENTED** [review]
Looks Good to me, other than the changes noted by Cameron and Dima.

File: ppr-ui/src/store/state/state-model.ts
Comment (garyfan1) [reply ppr-ui/src/store/state/state-model.ts:111]: The Api will return a 400 bad request if I pass in the wrong type. In the spec, it's asking for a string.
image

**Comment by garyfan1** [comment]
I was thinking, everyone was working according to the interface we had in our code, changing the types around can cause confusion and break previous code, e.g. `groupId` should be a `number` but was set to `string`, and lots of things break when I tried to change it to `string`. In the end, I map the type just for the Api call, maybe I can do the same thing for `phoneNumbers`?

**Comment by cameron-eyds** [comment]
> I was thinking, everyone was working according to the interface we had in our code, changing the types around can cause confusion and break previous code, i.e. `groupId` should be a `number` but was set to `string`, and lots of things break when I tried to change it to `string`. In the end, I map the type just for the Api call, maybe I can do the same thing for `phoneNumbers`?

phoneNumber is a string afaik, no need to map it. 
We should avoid mapping if we can, just fix our interfaces and values to adhere to the spec, if a discrepancy pops up.

File: ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts
Comment (garyfan1) [reply ppr-ui/src/interfaces/mhr-registration-interfaces/MhrRegistrationIF.ts]: removed!

File: ppr-ui/src/utils/mhr-api-helper.ts
Comment (garyfan1) [reply ppr-ui/src/utils/mhr-api-helper.ts]: it's actually redundant, removed.

File: ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue
Comment (garyfan1) [reply ppr-ui/src/views/newMhrRegistration/MhrRegistration.vue]: that's true, removed.

**Review by cameron-eyds — APPROVED** [review]
Thank you for the updates!

Approved pending a ticket to capture the tech debt from this PR (ie the Todo items). 
Please assign me to the ticket and post like to it here for visibility, thanks!

**Comment by garyfan1** [comment]
Ticket created for cleaning up the test code : bcgov/entity#13453

**Comment by cameron-eyds** [comment]
> Ticket created for cleaning up the test code : [bcgov/entity#13453](https://github.com/bcgov/entity/issues/13453)

Thank you!

**Review by dimak1 — APPROVED** [review]

**Pull request merged**
